pkg/expapi/types.go

```diff
@@ -27,3 +27,40 @@ support is experimental.
 */
 
 package expapi
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// SubresourceReference contains enough information to let you inspect or modify the referred subresource.
+type SubresourceReference struct {
+	Kind        string `json:"kind,omitempty"`
+	Namespace   string `json:"namespace,omitempty"`
+	Name        string `json:"name,omitempty"`
+	APIVersion  string `json:"apiVersion,omitempty"`
+	Subresource string `json:"subresource,omitempty"`
+}
+
+// HorizontalPodAutoscaler represents the configuration of a horizontal pod autoscaler.
+type HorizontalPodAutoscaler struct {
+	api.TypeMeta   `json:",inline"`
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// Target is a reference to Scale subresource. HorizontalPodAutoscaler will learn the current resource consumption from its status,
+	// and will set the desired number of pods by modyfying its spec.
+	Target *SubresourceReference `json:"target"`
+	// MinCount is the lower limit for the desired number of pods that can be set by the autoscaler.
+	MinCount int `json:"minCount"`
+	// MaxCount is the upper limit for the desired number of pods that can be set by the autoscaler. It cannot be smaller than MinCount.
+	MaxCount int `json:"maxCount"`
+	// TargetCPUConsumption is the target average CPU consumption of pods that the autoscaler will try to maintain by adjusting the desired number of pods.
+	TargetCPUConsumption float64 `json:"targetCPUConsumption"`
```

Comment:
Can we make this generic to resource name in the API?   Instead of something that is very locked in, can we make it opaque to type so that we don't have to migrate in the future?

For instance, make this a resource quantity that uses the CPU resource name and a field for percent?

    Target ResourceTarget
    type ResourceTarget struct {
      ResourceName string
      Percentage *int
    }

We have to reasons to prefer this - first, we have customers today who already expect to want to scale based on other metrics (heap usage of JVM is a good example), so we expect to get asked to add more fields for customers very rapidly.  Second, elsewhere in the API we don't code directly to resources - either in scheduling, limits, quota, or metrics.  So making this be generic (but for the first implementation rejecting anything other than cpu) prevents us from being locked in to this for future migrations.